sphinx/util/inspect.py

```diff
@@ -557,6 +557,45 @@ def signature_from_str(signature: str) -> inspect.Signature:
     return inspect.Signature(params, return_annotation=return_annotation)
 
 
+def signature_from_ast(node: ast.FunctionDef) -> inspect.Signature:
```

Comment:
I noticed there are duplicated codes in `signature_from_str()` and this. So it should be merged.